File: scripts/indicators.py
"""Pure-stdlib technical indicator calculations. No external dependencies."""


def sma(values, period):
    return sum(values[-period:]) / period


def ema_series(values, period):
    if len(values) < period:
        return []
    k = 2 / (period + 1)
    out = [sum(values[:period]) / period]
    for price in values[period:]:
        out.append(price * k + out[-1] * (1 - k))
    return out


def ema_last(values, period):
    series = ema_series(values, period)
    return series[-1] if series else None


def rsi(closes, period=14):
    if len(closes) < period + 1:
        return None
    gains, losses = [], []
    for i in range(1, len(closes)):
        change = closes[i] - closes[i - 1]
        gains.append(max(change, 0))
        losses.append(max(-change, 0))
    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period
    for i in range(period, len(gains)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def macd(closes, fast=12, slow=26, signal=9):
    if len(closes) < slow + signal:
        return None
    ema_fast = ema_series(closes, fast)
    ema_slow = ema_series(closes, slow)
    offset = len(ema_fast) - len(ema_slow)
    macd_line = [ema_fast[i + offset] - ema_slow[i] for i in range(len(ema_slow))]
    signal_line = ema_series(macd_line, signal)
    if not signal_line:
        return None
    hist = macd_line[-1] - signal_line[-1]
    return {"macd": macd_line[-1], "signal": signal_line[-1], "hist": hist}


def atr(highs, lows, closes, period=14):
    trs = []
    for i in range(1, len(closes)):
        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        trs.append(tr)
    if len(trs) < period:
        return None
    atr_val = sum(trs[:period]) / period
    for v in trs[period:]:
        atr_val = (atr_val * (period - 1) + v) / period
    return atr_val


def adx(highs, lows, closes, period=14):
    n = len(closes)
    if n < period * 2:
        return None
    plus_dm, minus_dm, trs = [], [], []
    for i in range(1, n):
        up_move = highs[i] - highs[i - 1]
        down_move = lows[i - 1] - lows[i]
        plus_dm.append(up_move if (up_move > down_move and up_move > 0) else 0)
        minus_dm.append(down_move if (down_move > up_move and down_move > 0) else 0)
        trs.append(
            max(
                highs[i] - lows[i],
                abs(highs[i] - closes[i - 1]),
                abs(lows[i] - closes[i - 1]),
            )
        )

    def wilder_smooth(vals):
        smoothed = [sum(vals[:period])]
        for v in vals[period:]:
            smoothed.append(smoothed[-1] - smoothed[-1] / period + v)
        return smoothed

    sm_plus = wilder_smooth(plus_dm)
    sm_minus = wilder_smooth(minus_dm)
    sm_tr = wilder_smooth(trs)
    plus_di = [100 * (sm_plus[i] / sm_tr[i]) if sm_tr[i] else 0 for i in range(len(sm_tr))]
    minus_di = [100 * (sm_minus[i] / sm_tr[i]) if sm_tr[i] else 0 for i in range(len(sm_tr))]
    dx = [
        100 * abs(plus_di[i] - minus_di[i]) / (plus_di[i] + minus_di[i])
        if (plus_di[i] + minus_di[i])
        else 0
        for i in range(len(plus_di))
    ]
    if len(dx) < period:
        return None
    adx_val = sum(dx[:period]) / period
    for v in dx[period:]:
        adx_val = (adx_val * (period - 1) + v) / period
    return {"adx": adx_val, "plus_di": plus_di[-1], "minus_di": minus_di[-1]}


def bollinger(closes, period=20, mult=2):
    if len(closes) < period:
        return None
    window = closes[-period:]
    mean = sum(window) / period
    variance = sum((c - mean) ** 2 for c in window) / period
    std = variance ** 0.5
    return {"upper": mean + mult * std, "middle": mean, "lower": mean - mult * std}

File: web/lib/indicators.ts
// Technical indicator calculations. Ported 1:1 from scripts/indicators.py.

export function sma(values: number[], period: number): number {
  const window = values.slice(-period);
  return window.reduce((a, b) => a + b, 0) / period;
}

export function emaSeries(values: number[], period: number): number[] {
  if (values.length < period) return [];
  const k = 2 / (period + 1);
  const seed = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
  const out = [seed];
  for (const price of values.slice(period)) {
    out.push(price * k + out[out.length - 1] * (1 - k));
  }
  return out;
}

export function emaLast(values: number[], period: number): number | null {
  const series = emaSeries(values, period);
  return series.length ? series[series.length - 1] : null;
}

export function rsi(closes: number[], period = 14): number | null {
  if (closes.length < period + 1) return null;
  const gains: number[] = [];
  const losses: number[] = [];
  for (let i = 1; i < closes.length; i++) {
    const change = closes[i] - closes[i - 1];
    gains.push(Math.max(change, 0));
    losses.push(Math.max(-change, 0));
  }
  let avgGain = gains.slice(0, period).reduce((a, b) => a + b, 0) / period;
  let avgLoss = losses.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (let i = period; i < gains.length; i++) {
    avgGain = (avgGain * (period - 1) + gains[i]) / period;
    avgLoss = (avgLoss * (period - 1) + losses[i]) / period;
  }
  if (avgLoss === 0) return 100.0;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export type Macd = { macd: number; signal: number; hist: number };

export function macd(closes: number[], fast = 12, slow = 26, signal = 9): Macd | null {
  if (closes.length < slow + signal) return null;
  const emaFast = emaSeries(closes, fast);
  const emaSlow = emaSeries(closes, slow);
  const offset = emaFast.length - emaSlow.length;
  const macdLine = emaSlow.map((_, i) => emaFast[i + offset] - emaSlow[i]);
  const signalLine = emaSeries(macdLine, signal);
  if (!signalLine.length) return null;
  const macdVal = macdLine[macdLine.length - 1];
  const signalVal = signalLine[signalLine.length - 1];
  return { macd: macdVal, signal: signalVal, hist: macdVal - signalVal };
}

export function atr(highs: number[], lows: number[], closes: number[], period = 14): number | null {
  const trs: number[] = [];
  for (let i = 1; i < closes.length; i++) {
    trs.push(
      Math.max(
        highs[i] - lows[i],
        Math.abs(highs[i] - closes[i - 1]),
        Math.abs(lows[i] - closes[i - 1])
      )
    );
  }
  if (trs.length < period) return null;
  let atrVal = trs.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (const v of trs.slice(period)) {
    atrVal = (atrVal * (period - 1) + v) / period;
  }
  return atrVal;
}

export type Adx = { adx: number; plus_di: number; minus_di: number };

export function adx(highs: number[], lows: number[], closes: number[], period = 14): Adx | null {
  const n = closes.length;
  if (n < period * 2) return null;
  const plusDm: number[] = [];
  const minusDm: number[] = [];
  const trs: number[] = [];
  for (let i = 1; i < n; i++) {
    const upMove = highs[i] - highs[i - 1];
    const downMove = lows[i - 1] - lows[i];
    plusDm.push(upMove > downMove && upMove > 0 ? upMove : 0);
    minusDm.push(downMove > upMove && downMove > 0 ? downMove : 0);
    trs.push(
      Math.max(
        highs[i] - lows[i],
        Math.abs(highs[i] - closes[i - 1]),
        Math.abs(lows[i] - closes[i - 1])
      )
    );
  }

  function wilderSmooth(vals: number[]): number[] {
    const smoothed = [vals.slice(0, period).reduce((a, b) => a + b, 0)];
    for (const v of vals.slice(period)) {
      smoothed.push(smoothed[smoothed.length - 1] - smoothed[smoothed.length - 1] / period + v);
    }
    return smoothed;
  }

  const smPlus = wilderSmooth(plusDm);
  const smMinus = wilderSmooth(minusDm);
  const smTr = wilderSmooth(trs);
  const plusDi = smTr.map((tr, i) => (tr ? (100 * smPlus[i]) / tr : 0));
  const minusDi = smTr.map((tr, i) => (tr ? (100 * smMinus[i]) / tr : 0));
  const dx = plusDi.map((p, i) => {
    const m = minusDi[i];
    return p + m ? (100 * Math.abs(p - m)) / (p + m) : 0;
  });
  if (dx.length < period) return null;
  let adxVal = dx.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (const v of dx.slice(period)) {
    adxVal = (adxVal * (period - 1) + v) / period;
  }
  return { adx: adxVal, plus_di: plusDi[plusDi.length - 1], minus_di: minusDi[minusDi.length - 1] };
}

export type Bollinger = { upper: number; middle: number; lower: number };

export function bollinger(closes: number[], period = 20, mult = 2): Bollinger | null {
  if (closes.length < period) return null;
  const window = closes.slice(-period);
  const mean = window.reduce((a, b) => a + b, 0) / period;
  const variance = window.reduce((a, c) => a + (c - mean) ** 2, 0) / period;
  const std = Math.sqrt(variance);
  return { upper: mean + mult * std, middle: mean, lower: mean - mult * std };
}
